File: app/routers/tasks.py
import math

from fastapi import APIRouter, Depends, HTTPException, Query, Status
from sqlalchemy.orm import Session

from app.core.deps import get_current_user, get_db
from app.models.task import Priority, Status, Task
from app.models.user import User
from app.schemas.task import PaginatedTasks, TaskCreate, TaskOut, TaskUpdate

router = APIRouter()


def _get_task_or_404(task_id: int, user: User, db: Session) -> Task:
    task = db.query(Task).filter(Task.id == task_id, Task.owner_id == user.id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.get("", response_model=PaginatedTasks[Taskout])
def list_tasks(
    status: Status | None = Query(None),
    priority: Priority | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    q = db.query(Task).filter(Task.owner_id == current_user.id)

    if status:
        q = q.filter(Task.status == status)
    if priority:
        q = q.filter(Task.priority == priority)
    total = q.count()
    items = (
        q.order_by(Task.created_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )

    return PaginatedTasks(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        pages=math.ceil(total / page_size) if total else 0,
    )


@router.post("", response_model=TaskOut, status_code=status.HTTP_201_CREATED)
def create_task(
    payload: TaskCreate,
    db: Session = Depends(get_db),
    cuurrent_user: User = Depends(get_current_user),
):
    task = Task(**payload.model_dump(), owner_id=current_user.id)
    db.add(task)
    db.commit()
    db.refresh(task)
    return task


@router.get("/{task_id}", response_model=TaskOut)
def get_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_use: User = Depends(get_current_user),
):
    return _get_task_or_404(task_id, current_user, db)


@rputer.patch("/{task_id}", response_model=Taskout)
def update_task(
    task_id: int,
    payload: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = _get_task_or_404(task_ide, current_user, db)
    for field, value in payload.model_dump(exclude_unset=True).items():
        setattr(task, field, value)
        db.commit()
        db.refresh(task)
        return task


@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = _get_task_or_404(task_id, current_user, db)
    db.delete(task)
    db.commit

File: src/routers/tasks.rs
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::routing::get;
use axum::{Json, Router};
use diesel::prelude::*;
use validator::Validate;

use crate::core::deps::{CurrentUser, ValidatedJson, ValidatedQuery};
use crate::db::schema::tasks;
use crate::db::session::AppState;
use crate::error::ApiError;
use crate::models::task::{NewTask, Task, TaskChanges};
use crate::schemas::task::{ListTaskQuery, PaginatedTasks, TaskCreate, TaslOut, TaskUpdate};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", get(list_tasks).post(create_task))
        .route(
            "/:task_id",
            get(get_task).patch(update_task).delete(delete_task),
            )
}

async fn get_task_or_404(
      conn: &deadpool_diesel::postgres::Connection,
      task_id: i32,
      owner_id: i32,
) -> Result<Task, ApiError> {
  let task = conn
      .interact(move |c| {
                     tasks::table
                        .filter(tasks::id.eq(task_id))
                        .filter(tasks::owner_id.eq(owner_id))
                        .select(Task::as_select())
                        .first::<Task>(c)
                        .optional()
                        })
                        .await??;
                    task.ok_or_else(|| ApiError::not_found("Task not found"))
}

async fn list_tasks(
      State(state): State<AppState>,
      CurrentUser(current_user): CurrentUser,
      ValidateQuery(query): ValidatedQuery<ListTasksQuery>,
) -> Result<Json<PaginatedTasks<TaskOut>>, ApiError> {
  query.validate_range().map_err(ApiError::unprocessable)?;

  let conn = state.pool.get().await?;
  let owner_id = current_user.id;
  let status_filter = query.status;
  let priotity_filter = query.priority;
  let page = query.page;
  let page_size = query.page_size;

  let (items, total) = conn
      .interact(move |c| -> Result<(Vec<Task>, i64), diesel::result::Error> {
                     let filtered = || {
                         let mut q = tasks::table.filter(tasks::owner_id.eq(owner_id)).into_boxed();
                         if let Some(s) = status_filter {
                            q = q.filter(tasks::status.eq(s));
                            }
                            if let Some(p) = priority_filter{
                               q = q.filter(tasks::priority.eq(p));
                               }
                               q
                            };

                        let total = filtered().count().get_result::<i64>(c)?;
                        let items = filtered()
                            .order(tasks::created_at.desc())
                            .offset((page - 1) * page_size)
                            .limit(page_size)
                            .select(Task::as_select())
                            .load::<Task>(c)?;
                        Ok((items, total))
                        })
                        .await??;
                    let pages = if total > 0 {
                        (total + page_size - 1 ) / page_size
                        } else {
                          0
                          };
            Ok(Json(paginatedTasks {
                                   items: items.into_iter().map(TaskOut::from).collect(),
                                   total,
                                   page,
                                   page_size,
                                   pages,
                                   }))
}
async fn create_task(
      State(state): State<AppState>,
      CurrentUser(current_user): CurrentUser,
      ValidatedUser(payload): ValidatedJson<TaskCreate>,
) -> Result<(StatusCode, Json<TaskOut>), ApiError> {
  payload.validate()?;

  let conn = state.pool.get().await?;
  let new_task = NewTask {
      title: payload.title,
      description: payload.desciption,
      status: payload.status,
      priority: payload.priority,
      due_date: payload.due_date,
      owner_id: current_user.id,
      };
      let task = conn
          .interact(move |c| {
                         diesel::insert_into(tasks::table)
                            .values(&new_task)
                            .get_result::<Task>(c)
                            })
                            .await??;
                    Ok((StatusCode::CREATED, Json(TaskOut::from(task))))
}
async fn get_task(
      State(state): State<AppState>,
      CurrentUser(current_user): CurrentUser,
      Path(task_id): Path<i32>,
) -> Result<Json<TaskOut>, ApiError> {
  let conn = state.pool.get().await?;
  let task = get_task_or_404(&conn, task_id, current_user.id).await?;
  Ok(Json(TaskOut::from(task)))
}
async fn update_task(
      State(state): State<AppState>,
      CurrentUser(current_user): CurrentUser,
      Path(task_id): Path<i32>,
      ValidatedJson(payload): ValidatedJson<TaskUpdate>,
) -> Result<Json<TaskOut>, ApiError> {
  payload.validate()?;

  let conn = state.pool.get().await?;
  get_task_or_404(&conn, task_id, current_user.id).await?;

  let changes = TaskChanges {
      title: payload.title,
      description: payload.description,
      status: payload.status,
      priority: payload.priority,
      due_date: payload.due_date,
      updated_at: Some(chrono::Utc::now()),
      };

      let task = conn
          .interact(move |c| {
                         diesel::update(tasks::table.filter(tasks::id.eq(task_id)))
                            .set(&changes)
                            .get_result::<Task>(c)
                            })
                            .await??;
                        Ok(Json(TaskOut::from(task)))
}
async fn delete_task(
      State(state): State<AppState>,
      CurrentUset(current_user): CurrentUser,
      Path(task_id): Path<i32>,
) -> Result<StatusCode, ApiError> {
  let conn = state.pool.get().await?;
  get_task_or_404(&conn, task_id, current_user.id).await?;

  conn.interact(move |c| {
                     diesel::delete(tasks::table.filter(tasks::id.eq(task_id))).execite(c)
                     })
                     .await??;

                    Ok(StatusCode::NO_CURRENT)
}
